mypy will never honor dependencies

The current mypy config currently ignores dependencies by (a) not installing them and then (b) [ignoring missing imports](https://github.com/pypa/twine/blob/c1807fac1275796cc8cbfbc29ce55596cd32ac80/mypy.ini#L21-L58) for those dependencies.

The non-installation of dependencies means that even for projects that have since added and declared typing support (both keyring and importlib_metadata fall into this boat), they're not tested.

I did find that if I repeated the dependencies in the type checks that the types appear to be checked, but that requires repeating the dependencies from setup.cfg (and thus keeping them in sync).  e.g.:

```
twine feature/importlib-metadata-entry-points-select $ git diff
diff --git a/mypy.ini b/mypy.ini
index dc7d662..ddc192f 100644
--- a/mypy.ini
+++ b/mypy.ini
@@ -22,10 +22,6 @@ strict_equality = True
 ; https://github.com/tartley/colorama/issues/206
 ignore_missing_imports = True
 
-[mypy-importlib_metadata]
-; https://github.com/python/importlib_metadata/issues/10
-ignore_missing_imports = True
-
 [mypy-keyring]
 ; https://github.com/jaraco/keyring/issues/437
 ignore_missing_imports = True
diff --git a/tox.ini b/tox.ini
index ae3bc0e..bf5d3ce 100644
--- a/tox.ini
+++ b/tox.ini
@@ -73,6 +73,7 @@ skip_install = True
 deps =
     mypy
     lxml
+    importlib_metadata >= 3.6
 commands =
     mypy --html-report mypy --txt-report mypy {posargs:twine}
     python -c 'with open("mypy/index.txt") as f: print(f.read())'
```

Is that the intention, to keep a separate subset of dependencies in the 'types' check?

In my other projects, I enable `ignore_missing_imports` globally and let dependencies declare and implement typing support at their pace. Any reason not to do that?

Yesterday, in keyring, I added [this change](https://github.com/jaraco/keyring/commit/a7dcac88f9dd3f4ecae492d1cbfe710262fc893f), which if I understand correctly should solve the failing type checks:

```
twine/auth.py:58: error: Call to untyped function "get_credential" in typed context
twine/auth.py:70: error: Call to untyped function "get_password" in typed context
```

But it didn't.

I'm surprised that mypy couldn't resolve the types for get_credential and get_password, because `get_keyring` resolves to a `KeyringBackend`, whose `get_credential` and `get_password` both declare return types.

In [this change](https://github.com/jaraco/keyring/commit/db6896acea942a86d3fbee7e2a556fffb38055ba), I even added types to the parameters for `get_password` and `get_credential`, but twine emits the same error.

I wonder if the problem is that the [accessor functions don't have type declarations](https://github.com/jaraco/keyring/blob/db6896acea942a86d3fbee7e2a556fffb38055ba/keyring/core.py#L54-L71). I would have expected mypy to infer the parameter and return types for those accessor functions from the types declared on the delegated behavior internally. I'm disinterested in copy/pasting types through every accessor function, and I'm not even sure that would help.

In jaraco/keyring@ceeaa92, I added the redundant type definitions I was reluctant to add. With these, twine now complains thus:

```
twine/auth.py:58: error: Argument 1 to "get_credential" has incompatible type "Optional[str]"; expected "str"
twine/auth.py:70: error: Argument 1 to "get_password" has incompatible type "Optional[str]"; expected "str"
twine/auth.py:70: error: Argument 2 to "get_password" has incompatible type "_lru_cache_wrapper[Optional[str]]"; expected "str"
```

--- a/mypy.ini
+++ b/mypy.ini
@@ -22,14 +22,6 @@ strict_equality = True
 ; https://github.com/tartley/colorama/issues/206
 ignore_missing_imports = True
 
-[mypy-importlib_metadata]
-; https://github.com/python/importlib_metadata/issues/10
-ignore_missing_imports = True
-
-[mypy-keyring]
-; https://github.com/jaraco/keyring/issues/437
-ignore_missing_imports = True
-
 [mypy-pkginfo]
 ; https://bugs.launchpad.net/pkginfo/+bug/1876591
 ignore_missing_imports = True

--- a/tox.ini
+++ b/tox.ini
@@ -72,6 +72,8 @@ commands =
 deps =
     mypy
     lxml
+    # required for more thorough type declarations
+    keyring >= 22.3
 commands =
     mypy --html-report mypy --txt-report mypy {posargs:twine}
     python -c 'with open("mypy/index.txt") as f: print(f.read())'

--- a/tox.ini
+++ b/tox.ini
@@ -69,7 +69,6 @@ commands =
     flake8 twine/ tests/
 
 [testenv:types]
-skip_install = True
 deps =
     mypy
     lxml

--- a/twine/auth.py
+++ b/twine/auth.py
@@ -55,7 +55,8 @@ class Resolver:
 
     def get_username_from_keyring(self) -> Optional[str]:
         try:
-            creds = keyring.get_credential(self.system, None)
+            system = cast(str, self.system)
+            creds = keyring.get_credential(system, None)
             if creds:
                 return cast(str, creds.username)
         except AttributeError:
@@ -67,7 +68,9 @@ class Resolver:
 
     def get_password_from_keyring(self) -> Optional[str]:
         try:
-            return cast(str, keyring.get_password(self.system, self.username))
+            system = cast(str, self.system)
+            username = cast(str, self.username)
+            return cast(str, keyring.get_password(system, username))
         except Exception as exc:
             warnings.warn(str(exc))
         return None

